### PR TITLE
[ADP-454] Runtime metrics

### PR DESCRIPTION
**Pull request by hasufell**
ADP-454

<details>
<summary>Teaser Prometheus</summary>

```
$ curl http://localhost:13798/metrics 
cardano_wallet_metrics_Stat_rtpriority_int 0
cardano_wallet_metrics_Stat_itrealvalue_int 0
rts_gc_par_max_bytes_copied 0
cardano_wallet_metrics_IO_syscr_int 54
cardano_wallet_metrics_Stat_minflt_int 1441
cardano_wallet_metrics_RTS_maxCompactBytes_B 0
cardano_wallet_metrics_Stat_cminflt_int 0
cardano_wallet_metrics_RTS_gcCopiedBytes_B 47608
rts_gc_num_gcs 221
cardano_wallet_metrics_Stat_enddata_int 29771744
cardano_wallet_metrics_Net_IpExt:OutOctets_B 1377542865
cardano_wallet_metrics_Stat_priority_int 20
cardano_wallet_metrics_IO_wchar_B 83
cardano_wallet_metrics_Stat_rsslim_int -1
cardano_wallet_metrics_RTS_cpuNs_ns 1920962
cardano_wallet_metrics_Stat_envend_int 140726935355228
cardano_wallet_metrics_Stat_pid_int 12809
cardano_wallet_metrics_Stat_ttynr_int 34823
cardano_wallet_metrics_Stat_startcode_int 4194304
rts_gc_mutator_wall_ms 9575
cardano_wallet_metrics_Stat_sigignore_int 0
rts_gc_gc_cpu_ms 105
cardano_wallet_metrics_Stat_cputicks_int 0
rts_gc_par_avg_bytes_copied 0
cardano_wallet_metrics_Stat_sigcatch_int 528390
cardano_wallet_metrics_Stat_majflt_int 0
cardano_wallet_metrics_Stat_cnswap_int 0
cardano_wallet_metrics_Stat_utime_int 0
cardano_wallet_metrics_Stat_blocked_int 0
cardano_wallet_metrics_Stat_cmajflt_int 0
rts_gc_init_cpu_ms 0
cardano_wallet_metrics_Stat_wchan_int 0
cardano_wallet_metrics_Stat_argend_int 140726935350409
cardano_wallet_metrics_IO_wbytes_B 0
cardano_wallet_metrics_Stat_startbrk_int 34394112
cardano_wallet_metrics_Stat_kstkeip_int 0
cardano_wallet_metrics_Stat_numthreads_int 6
cardano_wallet_metrics_Stat_session_int 26622
cardano_wallet_metrics_Stat_tpgid_int 12809
cardano_wallet_metrics_Mem_shared_int 5451
cardano_wallet_metrics_RTS_elapsedNs_ns 301678938
rts_gc_bytes_allocated 230364424
rts_gc_num_bytes_usage_samples 15
cardano_wallet_metrics_Mem_resident_int 6365
rts_gc_current_bytes_slop 227408
cardano_wallet_metrics_Stat_kstkesp_int 0
cardano_wallet_metrics_RTS_maxUsedMemBytes_B 2097152
cardano_wallet_metrics_Stat_argstart_int 140726935349989
cardano_wallet_metrics_Sys_Platform_int 1
rts_gc_max_bytes_slop 313624
cardano_wallet_metrics_IO_rchar_B 11062
rts_gc_par_tot_bytes_copied 0
cardano_wallet_metrics_Stat_flags_int 1077952512
cardano_wallet_metrics_Mem_data_int 11577
cardano_wallet_metrics_Stat_endcode_int 27238048
cardano_wallet_metrics_Mem_text_int 5626
cardano_wallet_metrics_RTS_gcElapsedNs_ns 82336
rts_gc_cumulative_bytes_used 68159136
cardano_wallet_metrics_Stat_rss_int 6365
cardano_wallet_metrics_Mem_size_int 268523351
ekg_server_timestamp_ms 1605554789360
cardano_wallet_metrics_Stat_cstime_int 0
cardano_wallet_metrics_RTS_maxLiveBytes_B 0
cardano_wallet_metrics_Stat_exitcode_int 0
cardano_wallet_metrics_Stat_guesttime_int 0
cardano_wallet_metrics_RTS_gcLiveBytes_B 133704
cardano_wallet_metrics_Stat_ppid_int 26622
rts_gc_gc_wall_ms 105
cardano_wallet_metrics_Stat_envstart_int 140726935350409
cardano_wallet_metrics_Stat_pgrp_int 12809
cardano_wallet_metrics_RTS_gcNum_int 1
rts_gc_mutator_cpu_ms 160
cardano_wallet_metrics_Stat_starttime_int 38375463
rts_gc_peak_megabytes_allocated 17
cardano_wallet_metrics_Stat_policy_int 0
cardano_wallet_metrics_Stat_startdata_int 27245632
cardano_wallet_metrics_IO_cxwbytes_B 0
cardano_wallet_metrics_Stat_vsize_int 1099871645696
cardano_wallet_metrics_RTS_maxSlopBytes_B 0
rts_gc_init_wall_ms 0
cardano_wallet_metrics_Stat_cutime_int 0
cardano_wallet_metrics_Net_IpExt:InOctets_B 12129589852
rts_gc_max_bytes_used 7362280
cardano_wallet_metrics_Stat_nswap_int 0
rts_gc_bytes_copied 27291728
cardano_wallet_metrics_Stat_blkio_int 0
cardano_wallet_metrics_IO_rbytes_B 0
cardano_wallet_metrics_Stat_startstack_int 140726935348176
cardano_wallet_metrics_RTS_maxLargeBytes_B 0
cardano_wallet_metrics_Stat_cguesttime_int 0
cardano_wallet_metrics_RTS_gcMajorNum_int 0
cardano_wallet_metrics_Sys_Pid_int 12809
rts_gc_wall_ms 9681
cardano_wallet_metrics_RTS_bytesAllocated_B 1129192
rts_gc_current_bytes_used 5765040
cardano_wallet_metrics_RTS_gcCpuNs_ns 82551
cardano_wallet_metrics_Stat_stime_int 0
cardano_wallet_metrics_Stat_signal_int 0
rts_gc_cpu_ms 266
cardano_wallet_metrics_Stat_exitsignal_int 17
cardano_wallet_metrics_Stat_processor_int 5
cardano_wallet_metrics_RTS_liveBytes_B 0
cardano_wallet_metrics_Stat_nice_int 0
cardano_wallet_metrics_IO_syscw_int 11
```

</details>

<details>
<summary>Teaser EKG</summary>

```
curl -H "Accept: application/json" http://localhost:13788/ | jq .
```

```json
{
  "iohk-monitoring version": {
    "type": "l",
    "val": "0.1.10.1"
  },
  "ekg": {
    "server_timestamp_ms": {
      "type": "c",
      "val": 1605553522731
    }
  },
  "rts": {
    "gc": {
      "gc_cpu_ms": {
        "type": "c",
        "val": 124
      },
      "mutator_wall_ms": {
        "type": "c",
        "val": 12178
      },
      "mutator_cpu_ms": {
        "type": "c",
        "val": 598
      },
      "gc_wall_ms": {
        "type": "c",
        "val": 118
      },
      "wall_ms": {
        "type": "c",
        "val": 12296
      },
      "bytes_copied": {
        "type": "c",
        "val": 60092216
      },
      "init_wall_ms": {
        "type": "c",
        "val": 0
      },
      "init_cpu_ms": {
        "type": "c",
        "val": 0
      },
      "max_bytes_used": {
        "type": "g",
        "val": 26806536
      },
      "max_bytes_slop": {
        "type": "g",
        "val": 1001208
      },
      "num_bytes_usage_samples": {
        "type": "c",
        "val": 12
      },
      "peak_megabytes_allocated": {
        "type": "g",
        "val": 40
      },
      "cpu_ms": {
        "type": "c",
        "val": 723
      },
      "current_bytes_used": {
        "type": "g",
        "val": 23334656
      },
      "bytes_allocated": {
        "type": "c",
        "val": 529295488
      },
      "par_max_bytes_copied": {
        "type": "g",
        "val": 0
      },
      "current_bytes_slop": {
        "type": "g",
        "val": 758016
      },
      "cumulative_bytes_used": {
        "type": "c",
        "val": 122968880
      },
      "num_gcs": {
        "type": "c",
        "val": 503
      },
      "par_tot_bytes_copied": {
        "type": "g",
        "val": 0
      },
      "par_avg_bytes_copied": {
        "type": "g",
        "val": 0
      }
    }
  },
  "cardano-wallet": {
    "wallet-engine": {
      "type": "l",
      "val": "\"\\\"9914f38a: local tip: 46f10b60<-[b40aa5a2-11184306#2045091]\\\"\""
    },
    "pools-engine": {
      "type": "l",
      "val": "\"\\\"Applying blocks [11184306 ... 11184306]\\\"\""
    },
    "metrics": {
      "Net": {
        "IpExt:InOctets": {
          "B": {
            "type": "g",
            "val": 12110694532
          }
        },
        "IpExt:OutOctets": {
          "B": {
            "type": "g",
            "val": 1374876709
          }
        }
      },
      "Mem": {
        "size": {
          "int": {
            "type": "g",
            "val": 268523646
          }
        },
        "text": {
          "int": {
            "type": "g",
            "val": 5901
          }
        },
        "data": {
          "int": {
            "type": "g",
            "val": 12365
          }
        },
        "resident": {
          "int": {
            "type": "g",
            "val": 7752
          }
        },
        "shared": {
          "int": {
            "type": "g",
            "val": 6347
          }
        }
      },
      "IO": {
        "rchar": {
          "B": {
            "type": "g",
            "val": 11062
          }
        },
        "syscw": {
          "int": {
            "type": "g",
            "val": 21
          }
        },
        "rbytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "wchar": {
          "B": {
            "type": "g",
            "val": 93
          }
        },
        "cxwbytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "syscr": {
          "int": {
            "type": "g",
            "val": 54
          }
        },
        "wbytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        }
      },
      "Sys": {
        "Platform": {
          "int": {
            "type": "g",
            "val": 1
          }
        },
        "Pid": {
          "int": {
            "type": "g",
            "val": 7172
          }
        }
      },
      "Stat": {
        "minflt": {
          "int": {
            "type": "g",
            "val": 2009
          }
        },
        "rtpriority": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "flags": {
          "int": {
            "type": "g",
            "val": 1077952512
          }
        },
        "endcode": {
          "int": {
            "type": "g",
            "val": 28361072
          }
        },
        "envend": {
          "int": {
            "type": "g",
            "val": 140736416722780
          }
        },
        "rsslim": {
          "int": {
            "type": "g",
            "val": -1
          }
        },
        "rss": {
          "int": {
            "type": "g",
            "val": 7752
          }
        },
        "priority": {
          "int": {
            "type": "g",
            "val": 20
          }
        },
        "guesttime": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "starttime": {
          "int": {
            "type": "g",
            "val": 38248539
          }
        },
        "cnswap": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "sigcatch": {
          "int": {
            "type": "g",
            "val": 544774
          }
        },
        "startdata": {
          "int": {
            "type": "g",
            "val": 28367904
          }
        },
        "majflt": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "cputicks": {
          "int": {
            "type": "g",
            "val": 2
          }
        },
        "numthreads": {
          "int": {
            "type": "g",
            "val": 6
          }
        },
        "blkio": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "nswap": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "vsize": {
          "int": {
            "type": "g",
            "val": 1099872854016
          }
        },
        "argend": {
          "int": {
            "type": "g",
            "val": 140736416717961
          }
        },
        "startbrk": {
          "int": {
            "type": "g",
            "val": 53985280
          }
        },
        "stime": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "processor": {
          "int": {
            "type": "g",
            "val": 7
          }
        },
        "argstart": {
          "int": {
            "type": "g",
            "val": 140736416717541
          }
        },
        "itrealvalue": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "cminflt": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "enddata": {
          "int": {
            "type": "g",
            "val": 30976544
          }
        },
        "ppid": {
          "int": {
            "type": "g",
            "val": 26622
          }
        },
        "sigignore": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "startcode": {
          "int": {
            "type": "g",
            "val": 4194304
          }
        },
        "envstart": {
          "int": {
            "type": "g",
            "val": 140736416717961
          }
        },
        "cstime": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "pgrp": {
          "int": {
            "type": "g",
            "val": 7172
          }
        },
        "exitcode": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "pid": {
          "int": {
            "type": "g",
            "val": 7172
          }
        },
        "ttynr": {
          "int": {
            "type": "g",
            "val": 34823
          }
        },
        "policy": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "utime": {
          "int": {
            "type": "g",
            "val": 2
          }
        },
        "blocked": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "cmajflt": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "cutime": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "wchan": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "kstkeip": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "startstack": {
          "int": {
            "type": "g",
            "val": 140736416715888
          }
        },
        "session": {
          "int": {
            "type": "g",
            "val": 26622
          }
        },
        "exitsignal": {
          "int": {
            "type": "g",
            "val": 17
          }
        },
        "tpgid": {
          "int": {
            "type": "g",
            "val": 7172
          }
        },
        "cguesttime": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "signal": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "kstkesp": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "nice": {
          "int": {
            "type": "g",
            "val": 0
          }
        }
      },
      "RTS": {
        "maxUsedMemBytes": {
          "B": {
            "type": "g",
            "val": 2097152
          }
        },
        "gcCpuNs": {
          "ns": {
            "type": "g",
            "val": 128700
          }
        },
        "maxCompactBytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "maxLargeBytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "bytesAllocated": {
          "B": {
            "type": "g",
            "val": 1129192
          }
        },
        "cpuNs": {
          "ns": {
            "type": "g",
            "val": 3184072
          }
        },
        "gcMajorNum": {
          "int": {
            "type": "g",
            "val": 0
          }
        },
        "gcCopiedBytes": {
          "B": {
            "type": "g",
            "val": 47608
          }
        },
        "elapsedNs": {
          "ns": {
            "type": "g",
            "val": 302668800
          }
        },
        "liveBytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "gcNum": {
          "int": {
            "type": "g",
            "val": 1
          }
        },
        "maxSlopBytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "gcElapsedNs": {
          "ns": {
            "type": "g",
            "val": 128435
          }
        },
        "maxLiveBytes": {
          "B": {
            "type": "g",
            "val": 0
          }
        },
        "gcLiveBytes": {
          "B": {
            "type": "g",
            "val": 133704
          }
        }
      }
    },
    "application": {
      "type": "l",
      "val": "\"\\\"Node is Haskell Node on testnet (1097911063).\\\"\""
    },
    "main": {
      "type": "l",
      "val": "\"\\\"Wallet backend server listening on 127.0.0.1:8090\\\"\""
    },
    "ntp-client": {
      "type": "l",
      "val": "\"\\\"local clock is drifting by -3214\\206\\188s\\\"\""
    },
    "network": {
      "type": "l",
      "val": "\"\\\"Query GetUpdateInterfaceState took 0.001308712s\\\"\""
    },
    "pools-db": {
      "type": "l",
      "val": "\"\\\"Will open db at ../cardano-node/run/wallet-state-test/stake-pools.sqlite\\\"\""
    },
    "wallet-db": {
      "type": "l",
      "val": "\"\\\"rnd_state_address table does not contain required field 'account_ix'. Adding this field with a default value of 2147483648.\\\"\""
    }
  }
}
```
</details>

### CODE REVIEW

File: lib/cli/src/Cardano/CLI.hs [hunk 1]
```diff
@@ -1538,13 +1557,39 @@ initTracer
 initTracer loggerName outputs = do
     cfg <- do
         c <- defaultConfigStdout
-        CM.setSetupBackends c [CM.KatipBK, CM.AggregationBK]
+        CM.setSetupBackends c [CM.KatipBK, CM.AggregationBK, CM.EKGViewBK, CM.EditorBK]
+        CM.setDefaultBackends c [CM.KatipBK, CM.EKGViewBK]
         CM.setSetupScribes c $ map mkScribe outputs
         CM.setDefaultScribes c $ map mkScribeId outputs
+        CM.setEKGBindAddr c $ Just (Endpoint ("127.0.0.1", 13788))
```
Comment (hasufell) [inline lib/cli/src/Cardano/CLI.hs]: I guess we want the ports be configurable... or alternatively, allow all of the logging config configurable and use yaml file like node does? I find yaml config a bit excessive and it is terribly documented anyway.

File: lib/cli/src/Cardano/CLI.hs [hunk 2]
```diff
@@ -1538,13 +1556,38 @@ initTracer
...
+        CM.setPrometheusBindAddr c $ Just ("127.0.0.1", 13798)
+        CM.setBackends c "cardano-wallet.metrics" (Just [CM.EKGViewBK])
         pure c
     (tr, sb) <- setupTrace_ cfg loggerName
+    EKG.plugin cfg tr sb >>= loadPlugin sb
+    startCapturingMetrics tr
     pure (sb, (cfg, tr))
 
+  where
+    startCapturingMetrics :: Trace IO Text -> IO ()
```
Comment (hasufell) [inline lib/cli/src/Cardano/CLI.hs:1656]: This is a nice copy paste of cardano-node.

**Comment by hasufell** [comment]
A general remark:

The EKG includes log messages, I believe the last of every tracer, or so. Prometheus doesn't seem to have picked up those or any metrics of internal warnings/errors. I don't really know if there's a way to add that.

**Comment by hasufell** [comment]
@disassembler can you check the [sample metrics](https://github.com/input-output-hk/cardano-wallet/pull/2322#issue-521870367) and whether they satisfy the acceptance criteria?

File: lib/cli/src/Cardano/CLI.hs (hunk 1 above)
Comment (Anviking) [reply lib/cli/src/Cardano/CLI.hs]: Yeah, we should probably add some CLI argument(s) in the style of
```
  --random-port            serve wallet API on any available port (conflicts
                           with --port)
  --port INT               port used for serving the wallet API. (default: 8090)
```

File: lib/cli/src/Cardano/CLI.hs (hunk 2 above)
Comment (Anviking) [reply lib/cli/src/Cardano/CLI.hs:1656]: A link to the original in a comment wouldn't hurt, such that we can compare them and see if the original change in the future, etc

**Review by Anviking — COMMENTED** [review]
lgtm, modulo
1. Should probably allow the port to be configurable, as you said.
2. I saw
```
[cardano-wallet.#ekgview.realizeFrom:Error:4] [2020-11-17 11:01:00.46 UTC] EKGView backend disabled due to initialisation error: EKGServerStartupError "Network.Socket.bind: resource busy (Address already in use)"
```
running integration tests. Maybe some node-wallet port conflict.

Also, how do you get the Prometheus output? Some (at least partial) steps for how to reproduce would be nice in the PR description. For EKG it was simple as I just had to visit `http://127.0.0.1:13788`, but not sure about Prometheus.

File: lib/cli/src/Cardano/CLI.hs (hunk 1 above)
Comment (hasufell) [reply lib/cli/src/Cardano/CLI.hs]: I think it's a rare enough use case to hide it behind env vars, which I just did. The test cluster than sets those env vars.

Comment (Anviking) [reply lib/cli/src/Cardano/CLI.hs]: > I think it's a rare enough use case to hide it behind env vars, which I just did.

Sounds sane 👍 

But: wouldn't it be nice if we extended the trace of the wallet url and port, to also include the EKG url/port (and maybe Prometheus) https://github.com/input-output-hk/cardano-wallet/blob/0c419518edf98cdb5f701fd6c65fe6832a89afad/lib/shelley/test/integration/Main.hs#L303 Otherwise it would be tricky to check the EKG of the integration tests wallet.

Comment (hasufell) [reply lib/cli/src/Cardano/CLI.hs]: Done

File: lib/shelley/test/integration/Main.hs [hunk 3]
```diff
@@ -168,21 +170,21 @@ main = withUtf8Encoding $ withTracers $ \tracers -> do
...
-                WalletsCLI.spec @n
-                HWWalletsCLI.spec @n
-                PortCLI.spec @t
-                NetworkCLI.spec @t
+                -- Hydra runs tests with code coverage enabled. CLI tests run
+                -- multiple processes. These processes can try to write to the
+                -- same .tix file simultaneously, causing errors.
+                --
+                -- Because of this, don't run the CLI tests in parallel in hydra.
+                parallelIf (not nix) $ describe "CLI Specifications" $ do
```
Comment (Anviking) [inline lib/shelley/test/integration/Main.hs]: The "CLI Specification" should not be inside "API Specifications". 

The `Settings.spec @n` indentation looks good though.

File: lib/cli/src/Cardano/CLI.hs [hunk 4]
```diff
@@ -1605,10 +1689,10 @@ withLogging =
 withLoggingNamed
     :: LoggerName
     -> [LogOutput]
-    -> ((CM.Configuration, Trace IO Text) -> IO a)
+    -> ((Switchboard Text, (CM.Configuration, Trace IO Text)) -> IO a)
```
Comment (Anviking) [inline lib/cli/src/Cardano/CLI.hs:1687]: It seems you changed this, but also currently ignore it?

File: lib/cli/src/Cardano/CLI.hs [hunk 5]
```diff
@@ -1576,26 +1599,87 @@ mkScribeId :: LogOutput -> ScribeId
...
+            uriRegName = ekg_host, uriPort = ekg_port
+        } } <- getEKGURL
     cfg <- do
         c <- defaultConfigStdout
-        CM.setSetupBackends c [CM.KatipBK, CM.AggregationBK]
+        CM.setSetupBackends c [CM.KatipBK, CM.AggregationBK, CM.EKGViewBK, CM.EditorBK]
+        CM.setDefaultBackends c [CM.KatipBK, CM.EKGViewBK]
         CM.setSetupScribes c $ map mkScribe outputs
         CM.setDefaultScribes c $ map mkScribeId outputs
+        CM.setEKGBindAddr c $ Just (Endpoint (ekg_host, read ekg_port))
```
Comment (Anviking) [inline lib/cli/src/Cardano/CLI.hs]: So you still enable EKG for both the wallet and the test tracer? How does that work?

Is it really the job of `initTracer` to read the `CARDANO_WALLET_EKG_PORT` env var? `initTracer` is also called by DB tests through `withLogging`.

Why complicate with `URI` types + parsing and formatting instead of using something more like `(String, Port)` or just `Port`, or `Int`?

How about _either_:
1. Take  two `Maybe Port` as arguments to `initTracer`. Could be put together with `outputs` inside a record if it seems better — `{ outputs, ekgPort, prometheusPort }`
2. Add `get{EKG,Prometheus}Port}FromEnv :: IO Port` to [Shelley.Launch](https://github.com/input-output-hk/cardano-wallet/blob/102fdb9e0052a76afee461f019969596c6efcf2f/lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs#L366-L376)

or 

Setup EKG and Prometheus using `setupEkgAndPrometheus :: Maybe Port -> Maybe Port -> CM.Configuration -> IO ()`, and calling from the integration main module, but I'm guessing you tried tried and it doesn't work?

File: lib/shelley/exe/cardano-wallet.hs [hunk 6]
```diff
@@ -293,14 +296,16 @@ withTracers
     -> (Trace IO MainLog -> Tracers IO -> IO a)
     -> IO a
 withTracers logOpt action =
-    withLogging [LogToStdout (loggingMinSeverity logOpt)] $ \(_, tr) -> do
+    withLogging [LogToStdout (loggingMinSeverity logOpt)] $ \(sb, (cfg, tr)) -> do
+        EKG.plugin cfg tr sb >>= loadPlugin sb
```
Comment (Anviking) [inline lib/shelley/exe/cardano-wallet.hs]: What does this do?

This does not affect the integration tests. And you don't have a similar call in the integration tests. Do both work as intended?

**Comment by piotr-iohk** [comment]
I think a description of how to leverege this feature (enable/disable, configure ports etc.) would be really handy. Perhaps a [wiki](https://github.com/input-output-hk/cardano-wallet/wiki) page?

Also AC says: "If no port is defined explicitely, there should be no metrics exposed." I think this is not the case and monitoring is always available on default ports even if custom ports are not provided.

File: lib/cli/src/Cardano/CLI.hs (hunk 4 above)
Comment (hasufell) [reply lib/cli/src/Cardano/CLI.hs:1687]: It is used here https://github.com/input-output-hk/cardano-wallet/pull/2322/files#diff-e12cf35dac52af31ad103700daedc7aea51fe3d2b10dbc97842e14e8c66bd8e9R299 and needed to start the snap server (ekg)

File: lib/cli/src/Cardano/CLI.hs (hunk 5 above)
Comment (hasufell) [reply lib/cli/src/Cardano/CLI.hs]: > So you still enable EKG for both the wallet and the test tracer? How does that work?

Nope, if you don't call `EKG.plugin cfg tr sb >>= loadPlugin sb`, then nothing will be exposed.

> Is it really the job of initTracer to read the CARDANO_WALLET_EKG_PORT env var? initTracer is also called by DB tests through withLogging.

I don't understand your concern here. The DB tests don't initialise the ekg plugin, but they are free to do so, if they wanted (e.g. to test logging, if that would be a thing).

> Why complicate with URI types + parsing and formatting instead of using something more like (String, Port) or just Port, or Int?

I don't have a strong opinion on this. So either is fine, I guess. Also feel free to provide a PR if you have detailed ideas on how to design this.

File: lib/shelley/exe/cardano-wallet.hs (hunk 6 above)
Comment (hasufell) [reply lib/shelley/exe/cardano-wallet.hs]: Yes, the integration tests don't need EKG/prometheus logging. That was the point of this change.

**Comment by hasufell** [comment]
> Also AC says: "If no port is defined explicitely, there should be no metrics exposed." I think this is not the case and monitoring is always available on default ports even if custom ports are not provided.

Good catch. Will adjust this.

File: lib/cli/src/Cardano/CLI.hs (hunk 5 above)
Comment (hasufell) [reply lib/cli/src/Cardano/CLI.hs]: I changed this quite a bit now

File: lib/cli/src/Cardano/CLI.hs [hunk 7]
```diff
@@ -1576,26 +1597,85 @@ mkScribeId :: LogOutput -> ScribeId
...
+                    pure Nothing
+        _ -> pure Nothing
+
+getEKGURL :: IO (Maybe (String, Port "EKG"))
+getEKGURL = do
+    ekg_port <- lookupEnv "CARDANO_WALLET_EKG_PORT"
+    ekg_host <- lookupEnv "CARDANO_WALLET_EKG_HOST"
+    case (ekg_host, ekg_port) of
+        (Just host, Just port) -> do
+            let parsedHost = readMaybe @String host
```
Comment (Anviking) [inline lib/cli/src/Cardano/CLI.hs]: `readMaybe @String` expects surrounding `""`, so I'd take you simply want `host`.

File: lib/shelley/exe/shelley-test-cluster.hs [hunk 8]
```diff
@@ -236,6 +236,8 @@ main = do
     whenReady tr trCluster db (RunningNode socketPath block0 (gp, vData)) = do
         let tracers = setupTracers (tracerSeverities (Just Info)) tr
         listen <- walletListenFromEnv
+        prometheusUrl <- (maybe "none" (\(h, p) -> T.pack h <> ":" <> toText @(Port "Prometheus") p)) <$> getPrometheusURL
+        ekgUrl <- (maybe "none" (\(h, p) -> T.pack h <> ":" <> toText @(Port "EKG") p)) <$> getEKGURL
```
Comment (Anviking) [inline lib/shelley/exe/shelley-test-cluster.hs]: These lines are longer than our style guide. In-particular, the important `get*URL` are way outside 80 chars.

**Review by Anviking — COMMENTED** [review]
> Yes, the integration tests don’t need EKG/prometheus logging. That was the point of this change.

You made it print e.g. `EKG url: 127.0.0.1:2000` in the integration startup, but that url doesn't actually lead anywhere. Let's allow EKG in integration tests too, particularly now when they're off by default?

> I don't understand your concern here. The DB tests don't initialise the ekg plugin, but they are free to do so, if they wanted (e.g. to test logging, if that would be a thing).

My concern is that unexpected things will happen when we build things on top of `initTracer`. without expecting it to use EKG. Perhaps `"Host or port value for EKB metrics invalid. Will be disabled."` will be printed where someone doesn't expect it. Maybe fine, but wouldn't have minded something else 🤔

**Comment by hasufell** [comment]
> You made it print e.g. EKG url: 127.0.0.1:2000 in the integration startup, but that url doesn't actually lead anywhere. Let's allow EKG in integration tests too, particularly now when they're off by default?

That's not really the case anymore. The URLs are not set when the env variables are not set and the tracer message now shows: `none`

> My concern is that unexpected things will happen when we build things on top of initTracer. 

I don't really see what. But I can start EKG/prometheus for the tests as well.

**Review by Anviking — COMMENTED** [review]
Running `cardano-wallet` and `shelley-test-cluster` I still see:
```
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] IO.ndisks = 0
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.interfaces = 13
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifn_"summary" = 0
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-mtu = 34644 B
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-ipackets = 91762790
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-ierrors = 0
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-opackets = 81394414
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-oerrors = 20179
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-collisions = 0
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-ibytes = 1137908736 B
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-obytes = 3096204288 B
[test-cluster.metrics:Notice:16] [2020-12-01 14:16:00.12 UTC] Net.ifd_0-imcast = 1004184
```
not sure why I don't see it in cardano-wallet:integration. Shouldn't they be hidden?

**Comment by hasufell** [comment]
> not sure why I don't see it in cardano-wallet:integration. Shouldn't they be hidden?

I don't know why we bother with enabling these for the tests even. They're not useful there. And you have to do it in every single executable.

**Comment by Anviking** [comment]
@hasufell I meant more that the `cardano-wallet` and `cardano-shelley-cluster` behaviour seemed wrong. All those notice logs at startup seem like unnecessary noise to stdout?

**Comment by hasufell** [comment]
should be fixed

**Review by Anviking — APPROVED** [review]
👌

**Comment by hasufell** [comment]
Bors r+

**Comment by piotr-iohk** [comment]
@hasufell I'm trying on testnet and have 2 questions:

**1. EKG `GC and memory statistics`** 

Starting like this:
```
CARDANO_WALLET_PROMETHEUS_HOST=6666 \
CARDANO_WALLET_EKG_PORT=7777 \
cardano-wallet serve --port 8090 \
  --node-socket ../relay1/node.socket \
  --testnet testnet-byron-genesis.json  \
  --database ./wallet-db 
```

I can see EKG metrics -> http://localhost:7777.
I can see Prometheus:
```
 $ curl http://localhost:6666/metrics 
cardano_wallet_metrics_Stat_rtpriority_int 0
...
```
But for EKG, there does not seem to be `GC and memory statistics` showing up/updating:

Screenshot from 2020-12-03 13-33-09
(Not sure if this is OK, or maybe sort of a config issue? Asking because I am able to see those metrics for the cardano-node when EKG is set up for it)

**2. Binding to something other than `localhost`**

There are those env variables:
```
CARDANO_WALLET_EKG_HOST
CARDANO_WALLET_PROMETHEUS_HOST
```

So I wanted to bind metrics to `0.0.0.0` or `127.0.0.1` but with that my wallet doesn't want to connect to the node :man_shrugging: 

```
CARDANO_WALLET_EKG_PORT=6666 \
CARDANO_WALLET_EKG_HOST=0.0.0.0 \
CARDANO_WALLET_PROMETHEUS_PORT=7777 \
CARDANO_WALLET_PROMETHEUS_HOST=0.0.0.0 \
cardano-wallet serve --port 8090 \
  --node-socket /path_to/cardano-node.socket \
  --testnet testnet-byron-genesis.json \
  --database ./wallet-db
...
[cardano-wallet.network:Notice:14] [2020-12-03 12:39:12.28 UTC] Couldn't connect to node (x2). Retrying in a bit...
[cardano-wallet.network:Notice:16] [2020-12-03 12:39:12.28 UTC] Couldn't connect to node (x2). Retrying in a bit...
[cardano-wallet.network:Notice:25] [2020-12-03 12:39:12.31 UTC] Couldn't connect to node (x2). Retrying in a bit...
[cardano-wallet.network:Warning:16] [2020-12-03 12:39:12.53 UTC] Couldn't connect to node (x3). Retrying in a bit..
```

Basically it seems that if anything is put in the `*_HOST` env variable the result is :point_up_2: .

**Comment by hasufell** [comment]
@piotr-iohk weird... I tried it and I cannot reproduce the issue with hostname

**Comment by KtorZ** [comment]
Note that EKG requires the `-T` runtime flag to work with monitoring https://hackage.haskell.org/package/ekg-0.4.0.15/docs/System-Remote-Monitoring.html

**Comment by piotr-iohk** [comment]
> @piotr-iohk weird... I tried it and I cannot reproduce the issue with hostname

Must have been something on my end. I cannot reproduce it either now.
